src/vs/platform/secrets/common/secrets.ts

```diff
@@ -0,0 +1,136 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { SequencerByKey } from 'vs/base/common/async';
+import { IEncryptionService } from 'vs/platform/encryption/common/encryptionService';
+import { createDecorator } from 'vs/platform/instantiation/common/instantiation';
+import { IStorageService, IStorageValueChangeEvent, StorageScope, StorageTarget } from 'vs/platform/storage/common/storage';
+import { Emitter, Event } from 'vs/base/common/event';
+import { ILogService } from 'vs/platform/log/common/log';
+
+export const ISecretStorageService = createDecorator<ISecretStorageService>('secretStorageService');
+
+export interface ISecretStorageProvider {
+	get(key: string): Promise<string | undefined>;
+	set(key: string, value: string): Promise<void>;
+	delete(key: string): Promise<void>;
+}
+
+export interface ISecretStorageService extends ISecretStorageProvider {
+	readonly _serviceBrand: undefined;
+	onDidChangeSecret: Event<string>;
+}
+
+export class SecretStorageService implements ISecretStorageService {
+	declare readonly _serviceBrand: undefined;
+
+	private _storagePrefix = 'secret://';
+
+	private readonly _onDidChangeSecret = new Emitter<string>();
+	onDidChangeSecret: Event<string> = this._onDidChangeSecret.event;
+
+	private readonly _sequencer = new SequencerByKey<string>();
+	private initialized = this.init();
+
+	private _basicStorageService: IBasicStorageService;
+
+	constructor(
+		@IStorageService storageService: IStorageService,
+		@IEncryptionService private _encryptionService: IEncryptionService,
+		@ILogService private readonly _logService: ILogService,
+	) {
+		this._basicStorageService = storageService;
+		this._basicStorageService.onDidChangeValue(e => this.onDidChangeValue(e.key));
```

Comment:
Just fyi, this event will also fire when you call `store` here in this file. If you want to avoid extra work then you need to suspend the listener.

Reply:
I would love to avoid the extra work. How can I suspend the listener? Any good patterns of doing that?

Reply:
Maybe pipe through a `PauseableEmitter`